features/layer_lock.c

```diff
@@ -76,6 +107,9 @@ void layer_lock_invert(uint8_t layer) {
     }
 #endif  // NO_ACTION_ONESHOT
     layer_on(layer);
+    #if LAYER_LOCK_IDLE_TIMEOUT > 0
+        layer_lock_timer = timer_read();
```

Comment:
It should be enough to restart the timer in `process_layer_lock()` on each key event. This line can be safely removed.

Reply:
Aha, got it, thanks for the detailed description.

> I think this is because the timer was only being updated if a layer was locked already, but I don't know if it's really necessary to limit it to update only when a layer is locked.

Maybe it was a matter of order? If the logic was like

```
if (any layer locked) { update timer }
if (keycode == lock_keycode && record->event.pressed) { lock the layer }
```

Then pressing the layer lock key would lock the current layer but without updating the timer. If the timer was from more than 32-ish seconds ago, it's possible the lock times out immediately.

I think the best thing to do to make the timeout simple and robust is to update the timer at the beginning of `process_layer_lock()` for every key event, even when no layer is locked.

> I was trying to reduce overhead I guess.

That's a fair concern. Fortunately, as far as I can tell, updating a timer is very cheap, so it's no a problem.

Reply:
I think you're right, the simplest solution is to update the timer always, regardless of whether a layer is locked or not. That should fix the issue, I think. I'll try making this change this evening and let you know how it goes. I'm still not really clear on why updating the timer inside the layer_lock function would have caused the issues I was seeing. I'll have to investigate that some more to satisfy my curiosity. I did try changing up the order as you suggest and I got the same behavior either way.

Reply:
I've moved the timer_read to the top of the process_layer_lock function and it works as expected there.